### DISCUSSION TITLE
"non-idempotent" req/rep

### DISCUSSION THREAD

**Discussion by hungrycat** [nanomsg/nng#1857] [Q&A] [answered]
I am seeking a bit of clarification on the "reliable" properties of the req protocol. I had originally read the [documentation](https://nng.nanomsg.org/man/v1.8.0/nng_req.7.html) of `NNG_OPT_REQ_RESENDTIME`
> Requests are also automatically resent if the peer to whom the original request was sent disconnects

as "re-send upon reconnection is independent of the `NNG_OPT_REQ_RESENDTIME` option". However, https://github.com/nanomsg/nng/issues/1300 seems to tie the two together.

So my question is: Is the `NNG_OPT_REQ_RESENDTIME` option controlling all aspects of the retry behaviour? With it set to infinite, will there be no scenario in which the protocol may decide to re-send a request, dropping therefore the requirement on the services being idempotent?

**Comment by gdamore** [comment, marked as answer]
You may in fact set the `NNG_OPT_REQ_RESENDTIME` to infinite, but an even better solution is to set it to 0.  Setting it to zero does in fact eliminate the need for the request to be idempotent, and there are some optimizations done for 0 that won't be done for infinite time.  (In particular, we won't bother setting a timer, etc.). So yes, recommendation is to to set to 0 for the behavior you want.
